File: find_previous_better/find_previous_better.py
def previousBest(n,ranks):
    stack = []
    previous_greater = []
    for i in range(n):
        while stack:
            if stack[-1] > ranks[i]:
                previous_greater.append(stack[-1])
                break

            else:
                stack.pop()

        if not stack:
            previous_greater.append(-1)

        stack.append(ranks[i])

    for i in previous_greater:
        print(i,end=" ")
        

if __name__=="__main__":
    n = int(input())
    ranks = list(map(int,input().rstrip().split()))

    previousBest(n,ranks)
    

File: find_previous_better/find_previous_better.java
import java.io.*;
import java.util.*;
import java.util.stream.*;
import static java.util.stream.Collectors.toList;


class PreviousBest{
    void previousBest(int n,int[] ranks){
        int[] stac = new int[n+1];
        int top = -1;
        int[] previous_greater = new int[n];
        for(int i=0;i<n;i++){
            while(top>-1){
                if(stac[top]>ranks[i]){
                    previous_greater[i] = stac[top];
                    break;
                }  
                else{
                    top--;
                }  
            }
            if(top<0){
                previous_greater[i] = -1;
            }
            stac[++top] = ranks[i];
            
        }

        for(int i=0;i<n;i++){
            System.out.print(previous_greater[i]+" ");
        }
    }
}

public class find_previous_better {
    public static void main(String[] args)throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine().trim());
        
        List<String> input = Stream.of(reader.readLine().split(" ")).collect(toList());

        int[] ranks = new int[n];
        for(int i=0;i<n;i++){
            ranks[i] = Integer.parseInt(input.get(i));
        }


        PreviousBest best = new PreviousBest();
        best.previousBest(n, ranks);
        
    }
    
}